### ISSUE TITLE
Garbage Collector implementation

### ISSUE THREAD

**Issue by Maaarcocr** [zetavm/zetavm#66]
TL;DR This is my attempt of implementing a garbage collector for ZetaVM, if you have any suggestions please comment.

In the next few weeks I'll try to implement a first version of a garbage collector for ZetaVM. 
Right now the VM is missing one, so it is relatively crucial to have one. I can't promise an incredibly good Garbage Collector, mainly because this is also my first Garbage Collector. It will be a very simple mark and sweep, with no moving memory and block the world.

(By Values with allocated memory I mean the Values that are created with [this](https://github.com/zetavm/zetavm/blob/c99411cca9d5c17dc6630b878f557cff249ef011/vm/runtime.cpp#L91) function)

The main point of my plan are:

- [ ] Implement a function (let's call this function get_stack_values for now) that returns all the Values with allocated memory on the stack at a given moment 
- [ ] Implement a LinkedList which will contain all the Values with allocated memory
- [ ] Implement the allocated function for the VM class.
- [ ] Add a check in the alloc (in the VM class) function that checks if a threshold value is overcome by the amount of allocated memory, if yes then starts the garbage collector function.
- [ ] Implement the garbage collector function. It will first call get_stack_values and mark the returned Values and their children recursively. It will then traverse the aforementioned LinkedList and sweep (free the memory) of the Values that have not been marked. It will then set the threshold value to double the size of the current allocated memory.

Tricky stuff that I should take care of (Please suggest more):

- Strings, they are in hashmap contained in the StringPool class, so I should also delete the entry from there

**Comment by maximecb** [comment]
There is one extra complication which you are probably not thinking of. That is, there can be some extra roots, heap-allocated values which are not referenced through the stack. The interpreter, the runtime, and functions in the core packages can refer to objects directly through `Value` instances, or through wrappers like `String`, `Object` and `Array` (which internally contain a `Value` instance).

In order to safely and correctly account for those roots, the GC has to be aware of them. One way to do this would be to have the `Value` objects form a doubly-linked list. Basically, every time you create a new `Value` object, it adds itself to a linked list. We can possibly save some time by avoiding doing this for values which are not heap-allocated. One tricky bit here is that copy and assignment constructors have to account for this correctly.

I wrote an implementation of something like this in Higgs:
https://github.com/higgsjs/Higgs/blob/master/source/runtime/gc.d#L60

Keeping track of every `Value` object will slow down Zeta, but we can get some speed back later by specializing functions like popInt64(), etc. to avoid creating `Value` objects. We can also add some unsafe operations to specialize some of the most common opcodes, but these optimizations should only be done once the GC has been thoroughly tested.

Besides this, some minor suggestions:

- We should probably rename the VM class to GC. That would better describe what that class does.

- You should probably use an `std::vector` instead of a linked list to keep track of stack values. linked lists make the most sense when you are really doing a lot of mutations, particularly removing and inserting elements in the middle of the list.

- Instead of plain mark and sweep, we could do lazy sweeping. This reduces pause times by doing the sweeping incrementally when allocating. The [notes found here](http://www.cs.tau.ac.il/~maon/teaching/2014-2015/seminar/seminar1415a-lec2-mark-sweep-mark-compact.pdf) explain this a bit.

**Comment by Maaarcocr** [comment]
> In order to safely and correctly account for those roots, the GC has to be aware of them. One way to do this would be to have the Value objects form a doubly-linked list.

This is what my second point in my list refers to, maybe I phrased it in a bad way.

**Comment by Maaarcocr** [comment]
About objects such as global_obj: 
Do we want them to be "global" and never be garbage collected or?

**Comment by maximecb** [comment]
> This is what my second point in my list refers to, maybe I phrased it in a bad way.

Just to make sure we're on the same page: there are two distinct sources of roots for the garbage collector. The stack that the interpreter uses to execute, and the `Value` objects that are created in various places in the Zeta program. You can't get at those by traversing the interpreter stacks. You can't call a function that will return a list of them. There has to be logic in the `Value` objects themselves that adds them to a list and removes them when they are destroyed.

> One thing I don't fully understand is how the VM will determine that the global_obj has to be marked or not. It doesn't seem to show up in the stack

The global object is referenced directly by the code. For instance, this is part of the code for some random function in the Plush language package:

```
block_239 = {
  instrs: [
    { op:'get_local', idx:0 },
    { op:'get_local', idx:1 },
    { op:'push', val:@global_obj },
    { op:'push', val:'rt_eq' },
    { op:'get_field' },
    { op:'call', ret_to:@block_241, num_args:2 },
  ]
};
```

The first `push` instruction has a reference to the global object, which it will push to the stack when executed. So the global object is not always on the stack, a reference to it only gets pushed to the stack when its value is needed.

What matters is that, functions, if they use the global object, will contain references to it in their code. So if a function is on the stack, and you go through the function during the mark phase, you will find references to the global object inside the function (if the function makes use of it). If there are no functions on the stack which make use of the global object, then that object is dead, and can be collected.

**Comment by Maaarcocr** [comment]
> Just to make sure we're on the same page: there are two distinct sources of roots for the garbage collector. The stack that the interpreter uses to execute, and the Value objects that are created in various places in the Zeta program. You can't get at those by traversing the interpreter stacks. You can't call a function that will return a list of them. There has to be logic in the Value objects themselves that adds them to a list and removes them when they are destroyed.

Yes, this is what I meant by: Implement a LinkedList which will contain all the Values with allocated memory. Every time a value such as an Object, Array or String is created it will get appended to the LinkedList. 

> What matters is that, functions, if they use the global object, will contain references to it in their code. So if a function is on the stack, and you go through the function during the mark phase, you will find references to the global object inside the function (if the function makes use of it). If there are no functions on the stack which make use of the global object, then that object is dead, and can be collected.

I can see this becoming quite slow, or maybe I'm wrong. If we have a very long call stack with very long function, going through all their instructions could be quite expensive. 

Also, do we want to allow the exported Values to be collected? (I may be misunderstanding this)

**Comment by maximecb** [comment]
> I can see this becoming quite slow, or maybe I'm wrong. If we have a very long call stack with very long function, going through all their instructions could be quite expensive.

It's kind of normal though. Every object that has been allocated has to be traversed. The only way to avoid that is a generational GC. Still, it might be surprisingly fast. Even with 100 million objects, we can do the traversal in 100 milliseconds or less. We can also make the mark phase multithreaded later on.

> Also, do we want to allow the exported Values to be collected? (I may be misunderstanding this)

No, we want to keep those alive. There should be no special handling necessary though. The VM keeps a reference to loaded packages here: https://github.com/zetavm/zetavm/blob/master/vm/packages.cpp#L467

So we have a `Value` object which holds a reference to the exports object of those packages. When the `Value` registers itself with the GC, that will keep the exports object alive.

**Comment by Maaarcocr** [comment]
> It's kind of normal though. Every object that has been allocated has to be traversed. The only way to avoid that is a generational GC. Still, it might be surprisingly fast. Even with 100 million objects, we can do the traversal in 100 milliseconds or less. We can also make the mark phase multithreaded later on.

I don't know why yesterday night I thought we had to go through the instructions as if they were still in a textual format. My bad, now it make more sense.

> No, we want to keep those alive. There should be no special handling necessary though. The VM keeps a reference to loaded packages here: https://github.com/zetavm/zetavm/blob/master/vm/packages.cpp#L467

So at every mark phase I would also mark the Values inside this vector, right? 

> Instead of plain mark and sweep, we could do lazy sweeping. This reduces pause times by doing the sweeping incrementally when allocating. The notes found here explain this a bit.

I will read the notes tonight. I may first try to implement a very basic mark and sweep as a learning experience and then change it to a lazy sweeping (if possible, or just restart the implementation).

**Comment by maximecb** [comment]
> So at every mark phase I would also mark the Values inside this vector, right?

If the `Value` objects add themselves to a linked list, then the GC can just mark all the roots in that list.

Take a look at the code I [linked to here](https://github.com/higgsjs/Higgs/blob/master/source/runtime/gc.d#L60). I would like `Value` to behave like `GCRoot` did in that code.

> I may first try to implement a very basic mark and sweep as a learning experience and then change it to a lazy sweeping (if possible, or just restart the implementation).

Sure, that's fine. The most important thing for the first version is that it works correctly :)

**Comment by maximecb** [comment]
I added some garbage collector tests I had written for the Higgs project to the repo: https://github.com/zetavm/zetavm/commit/c472a6c937376955836c049941eb1d8818a1bce3

We can port some of these to Plush in order to test your GC once it begins to work. As you can see in the code, I was relying on a few special instructions and runtime functions:

- `$rt_shrinkHeap` to shrink the size of the heap, and make collections more likely to happen

- `$ir_get_gc_count()` to get the number of collections that have happened so far

- `$ir_gc_collect(0)` to force a collection

Possibly, we could implement the equivalent of these in Zeta, say, as special functions inside a `core/vm/0` package.

We may also want to be able to enable special behaviors for debugging, such as zeroing out objects which are found dead, and setting pointers to dead objects to `false`. This would make it more likely for us to detect bugs. That is, if we collect objects that should not be collected, we will be able to tell. What is problematic with GC bugs is that they tend to go undetected for a while, so we have to be rigorous when it comes to GC testing.

**Comment by maximecb** [comment]
Hey @Maaarcocr, did you have a chance to get started on the GC? Hit any technical obstacles?

**Comment by Maaarcocr** [comment]
@maximecb I did! I've started my first point. I'm sorry for not being quick, but I'm attending a Microsoft competition for students and it's being a bit intense.

**Comment by maximecb** [comment]
@Maaarcocr No rush :)

**Comment by Maaarcocr** [comment]
@maximecb I'm actually hitting a block now. 
I'll describe the situation: 
So we start the program and the VM starts reading the bytecode, while doing so it creates the objects that represent the code. If the code is large, we may hit the GC, but nothing is on the stack yet. Can this happen?

**Comment by maximecb** [comment]
That shouldn't be a problem, because the VM interleaves code generation and execution, so it won't compile very much before execution is started.

Besides that, if we use `Value` objects in the VM to refer to GC'd values, then it doesn't matter if things are on the stack or not, because we're still keeping track of these references.

**Comment by Maaarcocr** [comment]
I just finished the get_stack_values function! I'm not sure what would be the best way of putting my code in the zeta repo. Right now most of the development is done in my fork and I'm okay with it. But if we want to also have a branch here that contains the GC code and to which I will make Pull Requests, so that I can have a better code review cycle, then I'm also okay with doing that.

**Comment by maximecb** [comment]
Good idea. I created the branch `gc-proto` which you can make PRs to. I recommend still pulling in changes from master frequently.

**Comment by Maaarcocr** [comment]
Okay perfect! Yes, I will rebase on master often!

**Comment by bigs** [comment]
just chiming in with a thought, but skip lists could be a wonderful aid in parallelizing mark passes!

**Comment by maximecb** [comment]
@bigs Because you can skip through the list? I guess it depends how many roots there are, but it might be possible to have a single thread copy out the roots into a vector, and then traverse those roots across multiple threads.

What I would also really like to have, eventually, is lazy, incremental sweeping. That could really help with pause times. That being said, we aren't even yet at the step where we have a GC ;)

**Comment by bigs** [comment]
basically using the skip lists as a means of dividing up the work. each thread gets an id (i) from [0..num threads) and finds its "batch of work" by traversing the second level skip list `i` times (looking forward one more step to know when it should jump back up to the second level), then dives in to the main linked list. it goes through and marks items until it hits the start of the next section (i.e. the next item in the second level skip list) at which point it goes back up to the second level and proceeds another `i` hops to find its next batch of work.

idk if that made any sense, but i'd be happy to elaborate. i'm sure there are far more intelligent approaches out there already though. i should probably do some reading!

do you have any recommended reading for GC?

**Comment by maximecb** [comment]
> do you have any recommended reading for GC?

Hmm. That's tricky because often, recent research papers will present fairly elaborate schemes.

You probably want to look at course notes to begin with, like these:
http://www.cs.cmu.edu/afs/cs/academic/class/15745-s06/web/handouts/garbage.pdf

More course notes:
http://www.math.grin.edu/~rebelsky/Courses/CS302/99S/Presentations/GC/

This blog post delves into some details of how the JVM does it:
http://www.pointsoftware.ch/en/under-the-hood-garbage-collection-part-1-different-object-life-cycles-short-and-long-term-requires-different-garbage-collection/

**Comment by SquidDev** [comment]
I've found the [LuaJIT wiki's algorithm section](http://wiki.luajit.org/New-Garbage-Collector#gc-algorithms) to be rather useful too. It's rather sparse on implementation details, but offers a decent description of some algorithms.

**Comment by bigs** [comment]
thanks y'all!
